### PR TITLE
Created Budget Simulation on the frontend

### PR DESCRIPTION
**Pull request by wilke0818**
- Created Citizen class that keeps boolean attributes for people
- Created a Population class that keeps a list of all of the Citizens
     - The number of Citizens is determined by a function parameter
     - Includes the will_support function that determines the number of citizens that will support a proposed budget
- Created a Serializer and functions in views.py to manage requests for data and to do the calculations
- Created a page on the frontend that contains the aggregate data for all of the citizens in the population that is generated (1000 people with attributes based on a distribution)
- Page also contains the ability to create a budget and submit it for approval from the population to see how many people support it
Screen Shot 2020-01-24 at 11 22 18 AM

### CODE REVIEW

File: backend/app/tests.py [hunk 1]
```diff
@@ -3,9 +3,162 @@
...
 from django.test import TestCase
+from .models import Population
+from .models import Citizen
 
 
 class MainTests(TestCase):
+    """
+    Tests for the models of Citizen and Population
+    """
     def test_is_this_on(self):
```
Comment (ryaanahmed) [inline backend/app/tests.py]: Go ahead and delete this trivial test

File: backend/app/tests.py [hunk 2]
```diff
@@ -3,9 +3,162 @@
...
+
+        cits_population = self.population_with_citizens
+        cits_population.create_citizens(5)
+
+        fake_cits_population = self.population_with_fake_citizens
+        # Creates a list of 5 dictionaries and then 5 Citizen objects
+        fake_cits_population.create_citizens(5)
+
+        # Test that the correct number of citizens were created
+        self.assertTrue(len(empty_population.citizen_list) == 0)
```
Comment (ryaanahmed) [inline backend/app/tests.py]: A bit clearer to do these with an `assertEqual`
`self.assertEqual(len(empty_population.citizen_list), 0)`

File: frontend/src/simulation/simulation.js [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+                        <div className="col-md-8 col-lg-5 col-sm-12 budget_container">
+                            <Budget
+                                population={this.state.population}
+                            />
+                        </div>
+                    </div>
+                </>
+            )
+        } else {
+            return (
```
Comment (ryaanahmed) [inline frontend/src/simulation/simulation.js]: I prefer to stick this kind of loading thing at the top of the render func --

```
if (!this.state.population) {
    return (
        <div>Loading!</div>
    )
}
```

-- bc then you can get on with the rest of the render function without having it all wrapped in a conditional

File: frontend/src/simulation/simulation.js [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+                </tbody>
+            </table>
+        )
+    }
+}
+AggregateData.propTypes = {
+    population: PropTypes.array,
+};
+
+class MainView extends React.Component {
```
Comment (ryaanahmed) [inline frontend/src/simulation/simulation.js]: Please rename this component into something descriptive of this whole mini-app. Remember it needs to live in a codebase with the other sims/games, too!

File: frontend/src/simulation/simulation.js [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+            categories: Object.keys(this.props.population[0]["traits"])
+        }
+    }
+
+    render() {
+        const aggregate_values = {};
+        for (let i = 0; i < this.state.categories.length; i++) {
+            let total = 0;
+            for (let j = 0; j < this.props.population.length; j++) {
+                total += this.props.population[j]["traits"][this.state.categories[i]]
```
Comment (ryaanahmed) [inline frontend/src/simulation/simulation.js]: factor out `this.state.categories[i]` and give it a name for clarity

File: frontend/src/simulation/simulation.js [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+        let sum = 0;
+        Object.keys(this.state.budgetProposal).forEach((resource) =>
+            sum += parseFloat(this.state.budgetProposal[resource]));
+        return sum <= 1
+    }
+
+
+    render() {
+        let result = 0;
+        if (this.state.reaction)
```
Comment (ryaanahmed) [inline frontend/src/simulation/simulation.js]: no braceless control flow, please! (I thought I'd disallowed this in the linter... will check now and make sure this is disallowed)

File: backend/config/urls.py [hunk 7]
```diff
@@ -19,19 +19,27 @@
...
     path('api/africa_map_geojson/', africa_map_geojson),
+    path('api/budget_response/', budget_response),
+    path('api/population/', population),
     path('api/democracy_scores/', democracy_score_json),
 
     # React views
     url('d3/', render_react_view, {'component_name': 'MapQuizD3'}),
     url('leaflet/', render_react_view, {'component_name': 'MapQuizLeaflet'}),
     url('svg/', render_react_view, {'component_name': 'MapQuizSVG'}),
+    url('simulation/', render_react_view, {'component_name': 'MainView'}),
```
Comment (ryaanahmed) [inline backend/config/urls.py]: better route name, please. `simulation` isn't descriptive -- everything in this project is going to be a simulation

**Review by ryaanahmed — CHANGES_REQUESTED** [review]
really great work. some changes requested.

One thing throughout the JS -- there are a _bunch_ of missing semicolons; please check. Our linter should have caught this, though, so I'll take a look at why it's not raising errors.

**Comment by ryaanahmed** [comment]
@montsegarza -- did you add real statistics? If so, please cite a source. Happy to merge once that's in.

**Comment by amy-lei** [comment]
@ryaanahmed I added an informal citation: 
`# Demographics from Afrobarometer 2016/2018 results `. 

Would this be enough?

**Comment by ryaanahmed** [comment]
@amy-lei thanks -- add a link, too, and it's okay for now -- particularly since a next step will be directly ingesting those stats in the backend. I'll add a formal citation.

**Comment by ryaanahmed** [comment]
I just fixed the linter semicolon thing -- eslint should now throw an error for statements that don't end in semicolons. Please fix, then I'll merge.

**Comment by amy-lei** [comment]
@ryaanahmed added the semicolons and link

**Comment by ryaanahmed** [comment]
@amy-lei THANK YOU! merging.

**Review by ryaanahmed — APPROVED** [review]

**Pull request merged**